src/track.c

```diff
@@ -796,19 +811,39 @@ s8 calc_navigation_measurement(u8 n_channels, const channel_measurement_t *meas[
                        &clock_err[i], &clock_rate_err[i]) != 0) {
       return -1;
     }
+  }
 
-    /* remove clock error to put all tots within the same time window */
-    if ((TOTs[i] + clock_err[i]) > min_TOF)
-      min_TOF = TOTs[i];
+  /* To calculate the pseudorange from the time of transmit we need the local
+   * time of reception. */
+  gps_time_t tor;
+  if (rec_time) {
+    /* If we were given a time, use that. */
+    tor = *rec_time;
+  } else {
+    /* If we were not given a time of reception then we can just set one of the
+     * pseudoranges aribtrarily to a nominal value and reference all the other
+     * pseudoranges to that. This doesn't affect the PVT solution but does
+     * potentially correspond to a large receiver clock error. */
+    tor = nav_meas[0]->tot;
+    tor.tow += GPS_NOMINAL_RANGE / GPS_C;
+    normalize_gps_time(&tor);
   }
 
   for (u8 i=0; i<n_channels; i++) {
-    nav_meas[i]->raw_pseudorange = (min_TOF - TOTs[i])*GPS_C + GPS_NOMINAL_RANGE;
-
-    nav_meas[i]->pseudorange = nav_meas[i]->raw_pseudorange \
-                               + clock_err[i]*GPS_C;
-    nav_meas[i]->doppler = nav_meas[i]->raw_doppler + clock_rate_err[i]*GPS_L1_HZ;
-
+    /* The raw pseudorange is just the time of flight divided by the speed of
+     * light. */
+    nav_meas[i]->raw_pseudorange = GPS_C * gpsdifftime(&tor, &nav_meas[i]->tot);
+
+    /* The corrected pseudorange, carrier_phase, Doppler applies the clock error
+     * and clock rate error correction from the ephemeris respectively. */
+    nav_meas[i]->pseudorange = nav_meas[i]->raw_pseudorange
+                               + clock_err[i] * GPS_C;
+    nav_meas[i]->carrier_phase = nav_meas[i]->raw_carrier_phase
```

Comment:
Both the carrier frequency and data are effected by satellite clock error (they are generated from the same clock).

Thats my understanding,
@anth-cole Is this correct?